compiler/src/dmd/expressionsem.d

```diff
@@ -11017,14 +11098,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 exp.e2 = exp.e2.implicitCastTo(sc, tb1next);
                 exp.type = tb1next.arrayOf();
             L2elem:
-                if (tb2.ty == Tarray || tb2.ty == Tsarray)
-                {
-                    // Make e2 into [e2]
-                    exp.e2 = new ArrayLiteralExp(exp.e2.loc, exp.type, exp.e2);
-                }
-                else if (checkNewEscape(sc, exp.e2, false))
+                if (!(tb2.ty == Tarray || tb2.ty == Tsarray) && checkNewEscape(sc, exp.e2, false))
```

Comment:
`checkNewEscape` shouldn't be disabled here. Formerly, the old branch changing `e2` into `[e2]` would have given an error when `e2` was a `scope` value. Now, the error would be along the lines of "`_d_arraycatnTX` not callable using `e2`" which is worse.

Reply:
I don't think I understand what you mean by disabling `checkNewEscape()`. The function is called if `e2` is not an array, just like before.

Regarding `scope`, I tried recreating the error you mentioned, but the following code compiles and runs just fine:
```d
struct S
{
    int x;

    this(this) { writeln("pblit: ", x); }
}

void main()
{
    scope S s1 = S(1);
    S[] s2 = [S(2), S(3)];

    auto s3 = s1 ~ s2;
    writeln(s3);
}
```
Is this what you meant?

Reply:
I mean this should fail because it puts x in allocated memory:
```D
int[][] f() @safe {
    scope int[] x;
    int[][] y;
    return y ~ x;
}
```

Weirdly, it doesn't error before this PR either because apparently the conversion from `y ~ x` to `y ~ [x]` doesn't check the array literal `[x]` for escaping pointers.

Reply:
So am I right to consider that removing `!(tb2.ty == Tarray || tb2.ty == Tsarray)` should do the trick? When I do this, your code prints the following message without `-dip1000` (coming from here [1]) and the first line turns into an error with `-dip1000`.

```
scope_arraycat.d(5): Deprecation: scope variable `x` may not be copied into allocated memory
scope_arraycat.d(5): Deprecation: scope variable `x` assigned to non-scope parameter `_param_1` calling `_d_arraycatnTX`
```

[1] https://github.com/dlang/dmd/blob/a4220358ecfcffe7ea38ab4a1996ffc5a5331f22/compiler/src/dmd/escape.d#L1047-L1062

Reply:
Yes, and don't do the lowering if checkNewEscape returns true so there's no redundant error

Reply:
Done @dkorpel. Unfortunately, `checkNewEscape()` only returns `true` when DIP 1000 is enabled, so I resorted to an ugly-ish special case in `functionParameters()` to skip checking `_d_arraycatnTX()`. Let me know if you think of a prettier way to do this.